optuna/storages/cached_storage.py

```diff
@@ -0,0 +1,328 @@
+                self._flush_trial(trial_id)
+                return True
+
+        return self._backend.set_trial_intermediate_value(trial_id, step, intermediate_value)
+
+    def set_trial_user_attr(self, trial_id: int, key: str, value: Any) -> None:
+
+        with self._lock:
+            cached_trial = self._get_cached_trial(trial_id)
+            if cached_trial is not None:
+                updates = self._get_updates(trial_id)
+                attrs = copy.copy(cached_trial.user_attrs)
+                attrs[key] = value
+                cached_trial.user_attrs = attrs
+                updates.user_attrs[key] = value
+                self._flush_trial(trial_id)
+                return
+
+        self._backend._update_trial(trial_id, user_attrs={key: value})
+
+    def set_trial_system_attr(self, trial_id: int, key: str, value: Any) -> None:
+
+        with self._lock:
+            cached_trial = self._get_cached_trial(trial_id)
+            if cached_trial is not None:
+                updates = self._get_updates(trial_id)
+                attrs = copy.copy(cached_trial.system_attrs)
+                attrs[key] = value
+                cached_trial.system_attrs = attrs
+                updates.system_attrs[key] = value
+                self._flush_trial(trial_id)
+                return
+
+        self._backend._update_trial(trial_id, system_attrs={key: value})
+
+    def _get_cached_trial(self, trial_id: int) -> Optional[FrozenTrial]:
+        if trial_id not in self._trial_id_to_study_id_and_number:
+            return None
+        study_id, number = self._trial_id_to_study_id_and_number[trial_id]
+        return self._studies[study_id].trials.get(number, None)
+
+    def _get_updates(self, trial_id: int) -> _TrialUpdates:
+        study_id, number = self._trial_id_to_study_id_and_number[trial_id]
+        updates = self._studies[study_id].updates.get(number, None)
+        if updates is not None:
+            return updates
+        updates = _TrialUpdates()
+        self._studies[study_id].updates[number] = updates
+        return updates
+
+    def get_trial(self, trial_id: int) -> FrozenTrial:
+
+        with self._lock:
+            trial = self._get_cached_trial(trial_id)
+            if trial is not None:
+                return copy.deepcopy(trial)
+
+        return self._backend.get_trial(trial_id)
+
+    def get_all_trials(self, study_id: int, deepcopy: bool = True) -> List[FrozenTrial]:
+
+        with self._lock:
+            trials = self._backend.get_all_trials(study_id, deepcopy=False)
+            if study_id in self._studies:
+                for key, trial in self._studies[study_id].trials.items():
+                    trials[trial.number] = trial
+            return copy.deepcopy(trials) if deepcopy else trials
+
+    def get_n_trials(self, study_id: int, state: Optional[TrialState] = None) -> int:
+
+        return self._backend.get_n_trials(study_id, state)
+
+    def flush(self) -> None:
```

Comment:
It seems that this method is not used in this PR. So how about removing it at the moment?
Or, if we keep this as a public method, I think that it's needed to add `with self._lock:` at the head of the body.

Reply:
Thank you. I removed the method.